design/cors-configuration.md

```diff
@@ -0,0 +1,188 @@
+
+At a high level the proposed changes will imply:
+- Adding new fields at route match level to configure the CORS policy in the YAML.
+- Changing some structs in the code.
+- Enabling the Envoy CORS filter.
+
+### Proposed YAML fields
+
+```yaml
+apiVersion: contour.heptio.com/v1beta1
+kind: IngressRoute
+metadata:
+ name: google
+ namespace: prod
+spec:
+ virtualhost:
+   fqdn: www.google.com
+   routes:
+   - match: /finance
+     # The CORS policy described here is local to match and apply for "/finance" only.
+     corsPolicy:
+       # Which domains can access the matched resources in a cross-site manner.
+       allowOrigin:
+         - "*"
+       # Which HTTP methods are allowed for cross-origin requests (case-sensitive).
+       allowMethods:
+         - GET
+         - POST
+       # The headers the server is going to accept (case-insensitive).
+       allowHeaders:
+         - cache-control
+         - content-type
+         - custom-header
+       # The non simple headers the client will be able to access (case-insensitive).
+       exposeHeaders:
+         - Content-Length
+         - Content-Range
+       # Whether the server allows sending credentials (cookies for instance) in cross-origin requests.
+       allowCredentials: true
+       # The number of seconds the preflight response can be cached.
+       maxAge: "1728000"
+     services:
+     - name: finance-app
+       port: 9999
+```
+
+The names and types of the proposed fields are inspired by the [HTTP headers](https://www.w3.org/TR/2010/WD-cors-20100727/#syntax). The `maxAge` is proposed as a string field type even though it's a delta-seconds. The motivation for this is that if we declare it as an `integer`, in cases where no value is provided, it  would get automatically the 0 default value. This is not what most people are going to expect and when not specified, it's better to apply the [browser defaults](https://developer.mozilla.org/en-US/docs/Web/HTTP/Headers/Access-Control-Max-Age#Directives).
+
+## Detailed Design
+
+### Reading from YAML
+The ingressroute spec struct will be updated to contain the CORS related information:
+
+```go
+// contour/apis/contour/v1beta1/ingressroute.go
+// Route contains the set of routes for a virtual host
+type Route struct {
+   [... other members ...]
+   // Specifies the CORS policy to apply for the matched path.
+    CorsPolicy *CorsPolicy `json:"corsPolicy,omitempty"`
+}
+
+// CorsPolicy allows for setting de CORS policy
+type CorsPolicy struct {
+    // Specifies whether the resource allows credentials.
+    AllowCredentials bool `json:"allowCredentials"`
+    // AllowOrigin specifies the origins that will be allowed to do CORS requests.
+    AllowOrigin []string `json:"allowOrigin"`
```

Comment:
`allow_origin` [has been deprecated](https://www.envoyproxy.io/docs/envoy/latest/version_history/v1.12.0.html?highlight=allow_origin#deprecated) in favor of `allow_origin_string_match` which is a [StringMatcher](https://github.com/envoyproxy/envoy/blob/549164c42cae84b59154ca4c36009e408aa10b52/api/envoy/type/matcher/string.proto#L20). I'm not sure about the type that would suit better here, the existing [MatchCondition](https://github.com/projectcontour/contour/blob/master/internal/dag/dag.go#L54) is somehow related...

Reply:
> I'm not sure about the type that would suit better here

I'd rather avoid to mimic Envoy's `StringMatcher` in Contour. I think it's too complicated and we would add too many parameters (exact, prefix, sufix, regex...). IMO defining our type as a String and translating it into an `exact` StringMatcher  would be good enough. I think this approach is reasonable because `*` [is always treated as a wildcard](https://github.com/envoyproxy/envoy/blob/d6e2fd0185ca620745479da2c43c0564eeaf35c5/source/extensions/filters/http/cors/cors_filter.cc#L142) in Envoy even for `exact` matchers. What does all this mean?

The following configuration would allow any origin:

```
virtualhost:
   fqdn: company.com
   corsPolicy:
     allowOrigin:
       - "*"

```

But if I wanted to allow any subdomain, I couldn't set that this way:

```
virtualhost:
   fqdn: company.com
   corsPolicy:
     allowOrigin:
       - "*.company.com"

```

I would need to set all the allowed subdomains explicitly instead:

```
virtualhost:
   fqdn: company.com
   corsPolicy:
     allowOrigin:
       - "foo.company.com"
       - "bar.company.com"
...
```